sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -954,7 +980,7 @@ object PushPredicateThroughJoin extends Rule[LogicalPlan] with PredicateHelper {
         split(splitConjunctivePredicates(filterCondition), left, right)
 
       joinType match {
-        case Inner =>
+        case _ if isInnerJoin(joinType, leftFilterConditions, rightFilterConditions) =>
```

Comment:
I don't think we should be doing this kind of reasoning in this rule.  If you are going to convert outer joins to inner joins that should be its own rule.